eclipse/src/saros/ui/wizards/AddProjectToSessionWizard.java

```diff
@@ -666,4 +680,79 @@ private void displaySaveDirtyEditorsDialog(final List<IEditorPart> dirtyEditors)
       for (IEditorPart editor : dirtyEditors) editor.doSave(new NullProgressMonitor());
     }
   }
+
+  private void storeCurrentProjectNameMapping(final JID jid) {
+    Map<String, String> currentProjectNameMapping = new HashMap<>();
+
+    for (final ProjectNegotiationData data : negotiation.getProjectNegotiationData()) {
+      final String projectID = data.getProjectID();
+      currentProjectNameMapping.put(
+          data.getProjectName(), namePage.getTargetProjectName(projectID));
+    }
+
+    setNewProjectNameMapping(jid, currentProjectNameMapping);
+  }
+
+  private Map<String, String> getLastProjectNameMapping(final JID jid) {
+    Map<String, String> result = new HashMap<>();
+
+    final String mapping = preferenceStore.getString(projectMappingStorePrefix + jid.getBase());
+
+    if (mapping == null || mapping.isEmpty()) return result;
+
+    String remoteName = null;
+    String localName = null;
+
+    for (int i = 0, j = 0; j < mapping.length(); j++) {
+
+      boolean isLastCharacter = j == mapping.length() - 1;
+
+      if (mapping.charAt(j) == ':' || isLastCharacter) {
+        if (j == 0) // corrupted data ...
+        return result;
+
+        if (!isLastCharacter && mapping.charAt(j - 1) == '\\') continue;
+
+        if (isLastCharacter) j++;
+
+        final String name = mapping.substring(i, j);
+
+        i = j + 1;
+
+        if (remoteName == null) remoteName = name;
+        else localName = name;
+
+        if (remoteName != null && localName != null) {
+          result.put(remoteName.replace("\\:", ":"), localName.replace("\\:", ":"));
+          remoteName = localName = null;
+        }
+      }
+    }
+
+    return result;
+  }
+
+  private void setNewProjectNameMapping(
+      final JID jid, final Map<String, String> projectNameMapping) {
+
+    final Map<String, String> lastProjectNameMapping = getLastProjectNameMapping(jid);
+
+    for (final Entry<String, String> entry : projectNameMapping.entrySet())
```

Comment:
Why not `lastProjectNameMapping.putAll(projectNameMapping)` (without loop)

Reply:
Because it was late when I wrote this  ....